cartography/intel/okta/oktaintel.py

```diff
@@ -0,0 +1,1225 @@
+    :param json_app_data: raw json group application assignment data.
+    :return: group ids as yield
+    """
+    app_data = json.loads(json_app_data)
+
+    for group in app_data:
+        yield group["id"]
+
+
+def _sync_okta_applications(neo4j_session, okta_org_id, okta_update_tag):
+    """
+    Sync okta application
+    :param neo4j_session: session from the Neo4j server
+    :param okta_org_id: okta organization id
+    :param okta_update_tag: The timestamp value to set our new Neo4j resources with
+    :return: Nothing
+    """
+    logger.debug("Syncing Okta Applications")
+
+    app_client = _create_application_client(okta_org_id)
+
+    data = _get_okta_applications(app_client)
+    _load_okta_applications(neo4j_session, okta_org_id, data, okta_update_tag)
+
+    api_client = _create_api_client(okta_org_id, "/api/v1/apps")
+
+    for app in data:
+        app_id = app["id"]
+        user_list = _get_application_assigned_users(api_client, app_id)
+        _ingest_application_user(neo4j_session, app_id, user_list, okta_update_tag)
+
+        group_list = _get_application_assigned_groups(api_client, app_id)
+        _ingest_application_group(neo4j_session, app_id, group_list, okta_update_tag)
+
+
+def _ingest_application_user(neo4j_session, app_id, user_list, okta_update_tag):
+    """
+    Add application users into the graph
+    :param neo4j_session: session with the Neo4j server
+    :param app_id: application to map
+    :param user_list: users to map
+    :param okta_update_tag: The timestamp value to set our new Neo4j resources with
+    :return: Nothing
+    """
+    ingest = """
+    MATCH (app:OktaApplication{id: {APP_ID}})
+    WITH app
+    UNWIND {USER_LIST} as user_id
+    MATCH (user:OktaUser{id: user_id})
+    WITH app, user
+    MERGE (user)-[r:APPLICATION]->(app)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {okta_update_tag}
+    """
+
+    neo4j_session.run(
+        ingest,
+        APP_ID=app_id,
+        USER_LIST=user_list,
+        okta_update_tag=okta_update_tag,
+    )
+
+
+def _ingest_application_group(neo4j_session, app_id, group_list, okta_update_tag):
+    """
+    Add application groups into the graph
+    :param neo4j_session: session with the Neo4j server
+    :param app_id: application to map
+    :param group_list: groups to map
+    :param okta_update_tag: The timestamp value to set our new Neo4j resources with
+    :return: Nothing
+    """
+    ingest = """
+    MATCH (app:OktaApplication{id: {APP_ID}})
+    WITH app
+    UNWIND {GROUP_LIST} as group_id
+    MATCH (group:OktaGroup{id: group_id})
+    WITH app, group
+    MERGE (group)-[r:APPLICATION]->(app)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {okta_update_tag}
+    """
+
+    neo4j_session.run(
+        ingest,
+        APP_ID=app_id,
+        GROUP_LIST=group_list,
+        okta_update_tag=okta_update_tag,
+    )
+
+
+def _get_user_id_from_graph(neo4j_session, okta_org_id):
```

Comment:
Same comment about using the `attach` pattern: https://github.com/lyft/cartography/blob/master/cartography/intel/aws/rds.py#L108

Reply:
Needed to fetch user factors and roles.